src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -143,15 +150,20 @@ public TextDocumentState UpdateText(SourceText newText, PreservationMode mode)
         return UpdateText(newTextAndVersion, mode);
     }
 
-    public TextDocumentState UpdateText(TextLoader loader, PreservationMode mode)
+    public TextDocumentState UpdateText(TextLoader? loader, PreservationMode mode)
     {
         // don't blow up on non-text documents.
-        var newTextSource = CreateTextFromLoader(loader, mode, this.SolutionServices);
+        var newTextSource = CreateTextAndVersionSource(SolutionServices, loader, FilePath, LoadTextOptions, mode);
 
         return UpdateText(newTextSource, mode, incremental: false);
     }
 
-    private static ITextAndVersionSource CreateTextFromLoader(TextLoader loader, PreservationMode mode, SolutionServices solutionServices)
+    protected static ITextAndVersionSource CreateTextAndVersionSource(SolutionServices solutionServices, TextLoader? loader, string? filePath, LoadTextOptions loadTextOptions, PreservationMode mode = PreservationMode.PreserveValue)
+        => loader != null
+            ? CreateTextFromLoader(solutionServices, loader, mode)
+            : CreateStrongText(TextAndVersion.Create(SourceText.From(string.Empty, encoding: null, loadTextOptions.ChecksumAlgorithm), VersionStamp.Default, filePath));
```

Comment:
this def needs docs.  it's unclear to me why we're being resilient to a null loader here.  couldn't we ask whoever is passing in null to instead pass in a loader that gives an empty string back?

Reply:
This hasn't changed. We accepted null loader before.

Reply:
k.  i hate it.  but can live with it.